From 44af1f5d806fb35f437a81e94370dedbdd8f538a Mon Sep 17 00:00:00 2001
From: Amir Vadai <avadai@redhat.com>
Date: Thu, 11 Sep 2014 12:01:23 -0400
Subject: [ethernet] mlx4: Keep only one driver entry release mlx4_priv

Message-id: <a1b0bfbeb57aab9dd7952012b6b04fe09461db08.1410425016.git.avadai@redhat.com>
Patchwork-id: 93738
O-Subject: [Patch RHEL7.1 V1 147/197] net/mlx4_core: Keep only one driver entry release mlx4_priv
Bugzilla: 1107617 1107618 1107619 1107620
RH-Acked-by: Jes Sorensen <Jes.Sorensen@redhat.com>

From: Wei Yang <weiyang@linux.vnet.ibm.com>

Following commit befdf89 "net/mlx4_core: Preserve pci_dev_data after
__mlx4_remove_one()", there are two mlx4 pci callbacks which will
attempt to release the mlx4_priv object -- .shutdown and .remove.

This leads to a use-after-free access to the already freed mlx4_priv
instance and trigger a "Kernel access of bad area" crash when both
.shutdown and .remove are called.

During reboot or kexec, .shutdown is called, with the VFs probed to
the host going through shutdown first and then the PF. Later, the PF
will trigger VFs' .remove since VFs still have driver attached.

Fix that by keeping only one driver entry which releases mlx4_priv.

Fixes: befdf89 ('net/mlx4_core: Preserve pci_dev_data after __mlx4_remove_one()')
CC: Bjorn Helgaas <bhelgaas@google.com>
Signed-off-by: Or Gerlitz <ogerlitz@mellanox.com>
Signed-off-by: Jack Morgenstein <jackm@dev.mellanox.co.il>
Signed-off-by: Wei Yang <weiyang@linux.vnet.ibm.com>
Signed-off-by: David S. Miller <davem@davemloft.net>
(cherry picked from commit da1de8dfff09d33d4a5345762c21b487028e25f5)
Signed-off-by: Amir Vadai <avadai@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/net/ethernet/mellanox/mlx4/main.c b/drivers/net/ethernet/mellanox/mlx4/main.c
index 0ab3ae1..18382be 100644
--- a/drivers/net/ethernet/mellanox/mlx4/main.c
+++ b/drivers/net/ethernet/mellanox/mlx4/main.c
@@ -2781,7 +2781,7 @@ static struct pci_driver mlx4_driver = {
  .name  = DRV_NAME,
  .id_table = mlx4_pci_table,
  .probe  = mlx4_init_one,
- .shutdown = mlx4_remove_one,
+ .shutdown = __mlx4_remove_one,
  .remove  = mlx4_remove_one,
  .err_handler    = &mlx4_err_handler,
 };
-- 
1.7.1